eclipse-theia/theia
# Multi window support for web views

#### What it does

This implements the multi window MVP proposal desribed in https://github.com/eclipse-theia/theia/issues/10526#issuecomment-1033950569

Support for moving webview-based views into a secondary window or tab.
For webview-based views a new button becomes available in the toolbar of the view to move the view to a secondary window.
This is only supported for webview-based views and only one view can be moved into the secondary window.
There can be multiple secondary windows though.

Primary code changes:
- Add concept of extractable widgets. Only widgets implementing the interface can be extracted.
- Add SecondaryWindowHandler that encapsulates logic to move widgets to new windows.
- Only webviews can be extracted
- Configure opened secondary windows in electron
  - Hide electron menu
  - Always use the native window frame for secondary windows to get window controls
  - Do not show secondary window icon if main window uses a custom title bar
- Contribute widget extraction button in a separate new extension `widget-extraction-ui`
- Extend application shell areas with a `secondaryWindow` area that contains all extracted widgets
- Extend frontend and webpack generators to generate the base html for secondary windows and copy it to the lib folder
- Bridge plugin communication securely via secure messaging between external webview and Theia:
  Webviews only accept messages from its direct parent window. This is necessary to avoid Cross Site Scripting (XSS).
  To make the messaging work in secondary windows, messages are sent to the external widget and then delegated to the webview's iframe.
  Thereby, the secondary window only accepts messages from its opener which is the theia main window.
  To achieve this, a webview knows the secondary window it is in (if any). It then sends messages to this window instead of directly to the webview iframe.
- Patch phosphor library during install via a postinstall hook. Remove check that widget attachment target must be in same DOM.

Contributed on behalf of ST Microelectronics and Ericsson and by ARM and EclipseSource.

Co-authored-by: Stefan Dirix <sdirix@eclipsesource.com>
Co-authored-by: robmor01 <Rob.Moran@arm.com>
Signed-off-by: Lucas Koehler <lkoehler@eclipsesource.com>

https://user-images.githubusercontent.com/6959840/164273386-8c032eac-72a3-4f97-ad7d-3e7966285bb7.mp4


#### Limitations

- Electron is unsupported. Generally, the functinality is existent but there are some issues with close handling of secondary windows.
- Only web views are supported. Plugin views don’t use webviews/iframes and thus cannot be extracted as part of this MVP.
- Widgets/editors that do not support moving from one area to another (e.g. main to bottom) also do not support moving to secondary windows
- If the last widget of an area is moved to a secondary window, it is still tracked as the former area’s current widget. This is caused by #10895 and also happens when moving widgets inside the main window
- When a webview in a secondary window opens a file picker - or possibly another dialog - it is opened in the main window instead of the secondary one.
- While an extracted webview resides in a secondary window, no contributed toolbar items contributed by the plugin are shown above the webview. This is caused by the absence of subsidiary UI (e.g. tabs, toolbar) in secondary windows. Fortunately, that isn't a very common use case since  webviews can integrate any additional UI necessary.
- External views cannot be restored to the main window, yet. To get a view back to the main window it needs to be closed and re-opened
- Windows are not restored on reload because they are not part of the stored application layout.
- extracted views do not respect "autoSave" (ex: for custom editors)

#### Screenshots
<details>
<summary><strong>Ubuntu (click to expand)</strong></summary>

![titlebar-ubuntu-custom-native](https://user-images.githubusercontent.com/6959840/164268781-4caa12e9-4e00-4b55-b554-6760820ae3a1.png)
![titlebar-ubuntu-native-native](https://user-images.githubusercontent.com/6959840/164268810-c6f72970-85bc-4418-bbfa-1704558b5b64.png)
</details>

<details>

<summary><strong>Windows 11 (click to expand)</strong></summary>

![titlebar-win11-custom-native-dark-no-icon](https://user-images.githubusercontent.com/6959840/164269558-ce00a427-c0d6-451f-abc3-463f659b8d8a.png)
![titlebar-win11-native-native](https://user-images.githubusercontent.com/6959840/164269598-d22b1b28-c8c6-4ade-a148-35c3481762d6.png)

</details>

<details>
<summary><strong>Mac (click to expand)</strong></summary>

![mac](https://user-images.githubusercontent.com/6959840/164269705-9b10e8d9-8425-4f8c-8ae2-fe76660de79e.png)
</details>

#### How to test

- Get one or more VS Code extensions that contribute web views and drop them in the `plugins` folder. For instance:
  - [Protein viewer](https://marketplace.visualstudio.com/items?itemName=ArianJamasb.protein-viewer), you can use protein `2btf` as example input
  - [Draw.io integration](https://marketplace.visualstudio.com/items?itemName=hediet.vscode-drawio)
  - [Mermaid Editor](https://marketplace.visualstudio.com/items?itemName=tomoyukim.vscode-mermaid-editor)
  - Build the webview sample providing a cat coding view from [here](https://github.com/microsoft/vscode-extension-samples/tree/main/webview-sample)
- If you use other extensions, make sure they support being moved around in the shell (e.g. from main to bottom) because webviews that don't support that won't support being moved to a new window either
- Open a web view and click the window button in its tool bar:
![how-to-move](https://user-images.githubusercontent.com/6959840/164268258-b929bef5-1e64-446a-9756-18e7c3100730.png)


#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
